### PR TITLE
add sol activate log

### PR DESCRIPTION
**Pull request by changev**
### **Foreword**

There're times that smoke-test failed for some strange reasons and it's hard to debug from RackHD log and Jenkins output. The ipmitool sol activate log can help to debug this tough test failure.
### **Details**
1. Need configure bmc user/password in build-config-setup jenkins job.
2. script `build-config` generate a `generate-sol-log.sh` from `generate-sol-log.sh.in`.
3. run `generate-sol-log.s` in virtualbox using `vagrant ssh -c` in `test.sh`
4. in pre-deploy.sh move sol logs to target folders for publishing. 

@panpan0000 @PengTian0 @jlongever @RackHD/corecommitters @cgx027

### CODE REVIEW

**Comment by changev** [comment]
test this please

**Comment by changev** [comment]
test this please

File: generate-sol-log.sh.in [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+                        cmd='ipmitool -I lanplus -H ${bmc_ip} -U ${bmc_user} -P ${bmc_password} sol activate |
+                                    tr 'A-Z' 'a-z' |
+                                    while IFS= read -r line; do 
+                                        echo "$(date) $line"; 
+                                    done > /home/vagrant/src/${vnode_num}.sol.log.raw'
+                        echo -e "bmc_ip=$ip\nbmc_user=${bmc%:*}\nbmc_password=${bmc#*:}\nvnode_num=${vnode_num}"|\
+                        sed 's/[\%]/\\&/g;s/\([^=]*\)=\(.*\)/s%${\1}%\2%/' > sed.script
+                        echo $cmd | sed -f sed.script > cmd.sh
+                        chmod oug+x cmd.sh
+                        screen -dmS sol bash cmd.sh
```
Comment (panpan0000) [inline generate-sol-log.sh.in:41]: comments can be nicer here.

**Comment by panpan0000** [comment]
@changev  showed me an example here :http://rackhdci.lss.emc.com/job/FIT/job/on-http/267/artifact/build/0.sol.log/*view*/
-  I was thinking change the log file name with MAC address of the Node, but I can still find MAC address in log like "52:54:be:ef:f6:49". so it's fine.
- I was thinking to use "idiscover" to find all BMCs, but @changev want to enable SOL until power state is on, so I think his implementation is fine.
- He explained to me it has to use “screen"  to run the "sol activate", simply run "sol activate" in background is not workable. (because the screen runs in Vagrant Box, so goes dead with Vagrant down, so it's ok.)

I think generally it's good to me now. :+1: 
**@RackHD/doccommitters ,** if you feel no problem, I would propose **merge this PR ASAP**. 
It's to capture more debug information in our random Jenkins errors , like "test_nodes_discovery"

**Comment by changev** [comment]
@panpan0000 Thanks! Such detailed review!!!

File: build-config [hunk 2]
```diff
@@ -99,3 +99,29 @@ if [ ! -z "${COMMIT}" ]; then
   fi
 fi
 
+echo "Generating sol-log script file from template"
+#BMC_{DOMAIN}_USER/BMC_{DOMAIN}_PASSWORD config in jenkins build-config-setup job parameters
+#This script generate a BMC_ACCOUNT_LIST="u1/p1 u2/p2 ..." according to these parameters.
+bmc_var_name_list=`env |grep BMC_ | awk -F '=' '{print $1}'|xargs`
```
Comment (cgx027) [inline build-config]: Code style: Suggest add white space before and after pipe '|'

File: build-config [hunk 3]
```diff
@@ -99,3 +99,29 @@ if [ ! -z "${COMMIT}" ]; then
...
+  if [[ "$domain_list" != *"$domain"* ]]; then
+    user_var_name=$domain"_USER"
+    password_var_name=$domain"_PASSWORD"
+    BMC_ACCOUNT_LIST="${!user_var_name}:${!password_var_name} $BMC_ACCOUNT_LIST"
+    domain_list="$domain $domain_list"
+  fi
+done
+export BMC_ACCOUNT_LIST
+if [ -e "${REPO_NAME}/generate-sol-log.sh.in" ]; then
+  env | sed 's/[\%]/\\&/g;s/\([^=]*\)=\(.*\)/s%${\1}%\2%/' > sed.script
```
Comment (cgx027) [inline build-config]: Maybe add some comments or examples to demo the regex function, for the sake of the reviewer and future maintainers :)

File: build-config [hunk 4]
```diff
@@ -99,3 +99,29 @@ if [ ! -z "${COMMIT}" ]; then
...
+  cat ${REPO_NAME}/generate-sol-log.sh.in | sed -f sed.script > generate-sol-log.sh
+  chmod oug+x generate-sol-log.sh
+  rm sed.script
+else
+  env | sed 's/[\%]/\\&/g;s/\([^=]*\)=\(.*\)/s%${\1}%\2%/' > sed.script
```
Comment (cgx027) [inline build-config]: Seems duplicate with line 118. Consider take it out of the if block if it will happen no matter if condition met or not.

**Review by cgx027 — COMMENTED** [review]
Looks good. :+1: after fix review comments.

**Review by panpan0000 — APPROVED** [review]

**Comment by panpan0000** [comment]
@RackHD/corecommitters , can you please merge it in if no issue ?
the Node Console Log is important to debug Jenkins instability issues.
Thanks

**Review by yyscamper — APPROVED** [review]
+1

**Pull request merged**